src/k8s/pkg/k8sd/app/hooks_join.go

```diff
@@ -253,5 +256,156 @@ func (a *App) onPostJoin(ctx context.Context, s state.State, initConfig map[stri
+	log.Info("Applying custom CRDs")
+	if err := k8sClient.ApplyCRDs(ctx, a.snap.K8sCRDDir()); err != nil {
+		log.Error(err, "Failed to apply custom CRDs")
+	}
+
+	if err := handleRollOutUpgrade(ctx, a.snap, s, k8sClient); err != nil {
+		return fmt.Errorf("failed to handle rollout-upgrade: %w", err)
+	}
+
 	return nil
 }
+
+func handleRollOutUpgrade(ctx context.Context, snap snap.Snap, s state.State, k8sClient *kubernetes.Client) error {
+	log := log.FromContext(ctx).WithValues("step", "rollout-upgrade")
+
+	log.Info("Checking if an upgrade is in progress")
+	upgrade, err := k8sClient.GetInProgressUpgrade(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to check for in-progress upgrade: %w", err)
+	}
+
+	thisNodeVersion, err := getNodeVersion(ctx, snap)
+	if err != nil {
+		return fmt.Errorf("failed to get node Kubernetes version: %w", err)
+	}
+
+	nodeVersions, err := k8sClient.NodeVersions(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get Kubernetes versions of cluster: %w", err)
+	}
+
+	if upgrade == nil {
+		return handleNoUpgradeInProgress(ctx, snap, s, k8sClient, thisNodeVersion, nodeVersions)
+	}
+
+	return handleUpgradeInProgress(ctx, s, k8sClient, upgrade, thisNodeVersion, nodeVersions)
+}
+
+func getNodeVersion(ctx context.Context, snap snap.Snap) (*versionutil.Version, error) {
+	versionStr, err := snap.NodeKubernetesVersion(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node Kubernetes version: %w", err)
+	}
+
+	version, err := versionutil.Parse(versionStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse node Kubernetes version %q: %w", versionStr, err)
+	}
+
+	return version, nil
+}
+
+func handleNoUpgradeInProgress(ctx context.Context, snap snap.Snap, s state.State, k8sClient *kubernetes.Client, thisNodeVersion *versionutil.Version, nodeVersions map[string]*versionutil.Version) error {
+	var clusterK8sVersion *versionutil.Version
+	for node, version := range nodeVersions {
+		if node == s.Name() {
+			continue
+		}
+		if clusterK8sVersion == nil {
+			clusterK8sVersion = version
+		}
+		if !clusterK8sVersion.EqualTo(version) {
+			return fmt.Errorf("the cluster has nodes with different Kubernetes versions %q and %q - upgrade all nodes to the same version before joining a new one", clusterK8sVersion, version)
+		}
+	}
+
+	if clusterK8sVersion == nil {
+		return fmt.Errorf("the cluster has no nodes - cannot determine cluster Kubernetes version")
+	}
+
+	if thisNodeVersion.EqualTo(clusterK8sVersion) {
+		return nil // Normal join
+	}
+
+	if thisNodeVersion.Major() == clusterK8sVersion.Major() && thisNodeVersion.Minor() == clusterK8sVersion.Minor() {
+		return fmt.Errorf("the joining node has a different patch version than cluster nodes")
+	}
+
+	if thisNodeVersion.Major() == clusterK8sVersion.Major() && thisNodeVersion.Minor() != clusterK8sVersion.Minor() {
+		return initiateRollingUpgrade(ctx, snap, s, k8sClient, thisNodeVersion, clusterK8sVersion)
+	}
+
+	return nil
```

Comment:
nit: Are we worried about major version changes?
```suggestion
	return initiateRollingUpgrade(ctx, snap, s, k8sClient, thisNodeVersion, clusterK8sVersion)
```